src/contracts/libraries/GPv2Encoding.sol

```diff
@@ -0,0 +1,119 @@
+// SPDX-license-identifier: LGPL-3.0-or-newer
+pragma solidity ^0.6.12;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+/// @title Gnosis Protocol v2 Encoding Library.
+/// @author Gnosis Developers
+library GPv2Encoding {
+    /// @dev A struct representing an executed order that can be used for
+    /// settling a batch.
+    struct Order {
+        address owner;
+        IERC20 sellToken;
+        IERC20 buyToken;
+        uint256 sellAmount;
+        uint256 buyAmount;
+        uint32 validTo;
+        uint256 tip;
+        OrderKind kind;
+        OrderFill fill;
+        uint256 executedAmount;
+    }
+
+    /// @dev An enum describing an order kind, either a buy or a sell order.
+    enum OrderKind {Sell, Buy}
+
+    /// @dev An enum describing an order's fill requirements.
+    enum OrderFill {Full, Partial}
+
+    /// @dev The stride of an encoded order.
+    uint256 internal constant ORDER_STRIDE = 204;
+
+    /// @dev Bit position for the [`OrderKind`] encoded in the flags.
+    uint8 internal constant ORDER_KIND_BIT = 0;
+    /// @dev Bit position for the [`OrderFill`] encoded in the flags.
+    uint8 internal constant ORDER_FILL_BIT = 1;
+
+    /// @dev Decodes a signed order from calldata bytes into memory.
+    /// Orders are tightly packed in order to reduce calldata and associated gas
+    /// costs. They contain the following fields:
+    /// ```
+    /// struct Order {
+    ///     uint8 sellTokenId;
+    ///     uint8 buyTokenId;
+    ///     uint256 sellAmount;
+    ///     uint256 buyAmount;
+    ///     uint32 validTo;
+    ///     uint256 tip;
+    ///     uint8 flags;
+    ///     uint256 executedAmount;
+    ///     uint32 nonce;
+    ///     Signature {
+    ///         uint8 v;
+    ///         bytes32 r;
+    ///         bytes32 s;
+    ///     } signature;
+    /// }
+    /// ```
+    /// @param tokens The list of tokens included in the settlement. Token IDs
+    /// represent indices in this array to actual addresses.
+    /// @param encodedOrder The order as encoded calldata bytes.
+    /// @param order The memory location to decode the order to.
```

Comment:
Why did you decide for handing over a memory location? Intuitively, I would have created a new memory object in the function call and then returned it.

Reply:
The reason is that Solidity doesn't have a concept of "garbage collection" or freeing memory. This would mean every function call would allocate a new `Order` that would then get copied into the main `Order` memory array, using twice as much memory as needed.